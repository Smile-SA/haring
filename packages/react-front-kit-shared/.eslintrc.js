module.exports = {
  extends: ['custom'],
  parserOptions: {
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
  },
  root: true,
};
