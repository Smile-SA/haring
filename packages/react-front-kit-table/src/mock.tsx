/* eslint-disable react-refresh/only-export-components */
export * from './Components/Table/Table.mock';
export * from './Components/TableGridView/TableGridView.mock';
