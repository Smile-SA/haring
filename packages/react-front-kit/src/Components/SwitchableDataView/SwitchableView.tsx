import type {
  SegmentedControlItem,
  SegmentedControlProps,
} from '@mantine/core/lib/SegmentedControl/SegmentedControl';
import type { ReactElement, ReactNode } from 'react';

import { Paper, SegmentedControl } from '@mantine/core';
import { useUncontrolled } from '@mantine/hooks';
import { createStyles } from '@mantine/styles';

const useStyles = createStyles(() => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    gap: 42,
    padding: 24,
  },
  content: {
    marginBottom: 48,
  },
  switchButton: {
    lineHeight: 0,
    padding: '0.5rem',
  },
  topBar: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 8,
  },
  topBarLeft: {
    margin: 'auto auto auto 0',
  },
  topBarRight: {
    marginLeft: 'auto 0 auto auto',
  },
}));

export interface IDataView extends SegmentedControlItem {
  dataView: ReactNode;
}

interface ISwitchableViewProps extends Omit<SegmentedControlProps, 'data'> {
  activeViewIndex?: number;
  defaultViewIndex?: number;
  onViewChange?: (index: number) => void;
  topBarLeft?: ReactNode;
  views: IDataView[];
}

export function SwitchableView(props: ISwitchableViewProps): ReactElement {
  const {
    activeViewIndex,
    defaultViewIndex,
    onViewChange,
    topBarLeft,
    views = [],
    ...segmentedControlProps
  } = props;
  const [activeViewState, handleChangeActiveView] = useUncontrolled<IDataView>({
    defaultValue: defaultViewIndex ? views[defaultViewIndex] : undefined,
    finalValue: views[0],
    onChange: (v) =>
      onViewChange
        ? onViewChange(views.findIndex((view) => view.value === v.value))
        : undefined,
    value: activeViewIndex ? views[activeViewIndex] : undefined,
  });
  const { classes } = useStyles();

  return (
    <Paper className={classes.container}>
      <div className={classes.topBar}>
        {Boolean(topBarLeft) && (
          <span className={classes.topBarLeft}>{topBarLeft}</span>
        )}
        <span className={classes.topBarRight}>
          <SegmentedControl
            classNames={{ label: classes.switchButton }}
            data={views}
            onChange={(v) =>
              handleChangeActiveView(
                views.find((view) => view.value === v) ?? views[0],
              )
            }
            radius={4}
            size="md"
            value={activeViewState.value}
            {...segmentedControlProps}
          />
        </span>
      </div>
      <div className={classes.content}>{activeViewState.dataView ?? ''}</div>
    </Paper>
  );
}
