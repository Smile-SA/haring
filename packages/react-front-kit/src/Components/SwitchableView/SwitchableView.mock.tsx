import { Center } from '@mantine/core';
import { ListBullets, SquaresFour, Star } from '@phosphor-icons/react';

export const viewMockA = (
  <Center
    style={{
      backgroundColor: 'lightgray',
      height: 200,
    }}
  >
    View A
  </Center>
);

export const viewMockB = (
  <Center
    style={{
      backgroundColor: 'lightblue',
      height: 200,
    }}
  >
    View B
  </Center>
);

export const viewMockC = (
  <Center
    style={{
      backgroundColor: 'lightpink',
      height: 200,
    }}
  >
    View C
  </Center>
);

export const viewsMock = [
  {
    dataView: viewMockA,
    label: <ListBullets />,
    value: 'a',
  },
  {
    dataView: viewMockB,
    label: <SquaresFour />,
    value: 'b',
  },
  {
    dataView: viewMockC,
    label: <Star />,
    value: 'c',
  },
];
