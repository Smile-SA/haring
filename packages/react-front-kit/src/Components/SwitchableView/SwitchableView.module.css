.container {
  display: flex;
  flex-direction: column;
  gap: 42px;
  padding: 24px;
}

.switchButton {
  line-height: 0px;
  padding: 0.5rem;
}

.topBar {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-top: 8px;
}

.topBarLeft {
  margin: auto auto auto 0;
  width: 100%;
}

.topBarRight {
  align-items: center;
  display: flex;
  flex-direction: row;
  gap: 16px;
  margin: auto 0 auto auto;
}
