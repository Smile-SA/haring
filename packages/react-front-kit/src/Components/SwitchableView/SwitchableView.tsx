import type { PaperProps } from '@mantine/core';
import type {
  SegmentedControlItem,
  SegmentedControlProps,
} from '@mantine/core/lib/SegmentedControl/SegmentedControl';
import type { ReactElement, ReactNode } from 'react';

import { Paper, SegmentedControl } from '@mantine/core';
import { useUncontrolled } from '@mantine/hooks';
import { createStyles } from '@mantine/styles';

const useStyles = createStyles(() => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    gap: 42,
    padding: 24,
  },
  content: {
    marginBottom: 48,
  },
  switchButton: {
    lineHeight: 0,
    padding: '0.5rem',
  },
  topBar: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 8,
  },
  topBarLeft: {
    margin: 'auto auto auto 0',
  },
  topBarRight: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: 'row',
    gap: 16,
    margin: 'auto 0 auto auto',
  },
}));

export interface IDataView extends SegmentedControlItem {
  dataView: ReactNode;
}

interface ISwitchableViewProps extends PaperProps {
  /** Default index of the active view */
  defaultValue?: number;
  /** Called when active view changes */
  onChange?: (index: number) => void;
  /** Extra Props, or function returning props for the SegmentedControl component, except 'data', 'defaultValue', 'onChange' and 'value' which are handled by this component */
  segmentedControlProps?: Omit<
    SegmentedControlProps,
    'data' | 'defaultValue' | 'onChange' | 'value'
  >;
  /** ReactNode for the left section of the top bar */
  topBarLeft?: ReactNode;
  /** ReactNode for the right section of the top bar, to the left of the SegmentedControl buttons */
  topBarRight?: ReactNode;
  /** Index of the active view, when component is controlled */
  value?: number;
  /** Array of all views */
  views: IDataView[];
}

/** Additional props will be forwarded to the [Mantine Paper component](https://mantine.dev/core/paper) */
export function SwitchableView(props: ISwitchableViewProps): ReactElement {
  const {
    defaultValue,
    onChange,
    segmentedControlProps,
    topBarLeft,
    topBarRight,
    value,
    views = [],
    ...paperProps
  } = props;
  const [activeViewIndex, setActiveViewIndex] = useUncontrolled<number>({
    defaultValue,
    finalValue: 0,
    onChange,
    value,
  });
  const activeView = views[activeViewIndex];
  const { classes } = useStyles();

  function handleViewChange(value: string): void {
    setActiveViewIndex(views.findIndex((view) => view.value === value));
  }

  return (
    <Paper className={classes.container} {...paperProps}>
      <div className={classes.topBar}>
        {Boolean(topBarLeft) && (
          <span className={classes.topBarLeft}>{topBarLeft}</span>
        )}
        <span className={classes.topBarRight}>
          {topBarRight}
          <SegmentedControl
            classNames={{ label: classes.switchButton }}
            data={views}
            onChange={handleViewChange}
            radius={4}
            size="md"
            value={activeView.value}
            {...segmentedControlProps}
          />
        </span>
      </div>
      <div className={classes.content}>{activeView.dataView ?? ''}</div>
    </Paper>
  );
}
