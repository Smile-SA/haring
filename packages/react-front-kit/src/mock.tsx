/* eslint-disable react-refresh/only-export-components */

export * from './Components/ResponsiveTabs/ResponsiveTabs.mock';
export * from './Components/SidebarMenu/SidebarMenu.mock';
export * from './Components/SwitchableView/SwitchableView.mock';
export * from './Components/Thumbnail/Thumbnail.mock';
export * from './Components/ThumbnailGrid/ThumbnailGrid.mock';
export * from './Layouts/FoldableColumnLayout/FoldableColumnLayout.mock';
