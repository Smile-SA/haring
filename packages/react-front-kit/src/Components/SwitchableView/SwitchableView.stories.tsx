import type { Meta, StoryObj } from '@storybook/react';

import { SwitchableView as Cmp } from './SwitchableView';
import { views } from './SwitchableView.mock';

const meta = {
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/SwitchableView',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const SwitchableView: IStory = {
  args: {
    topBarLeft: <span>TopBarLeft...</span>,
    topBarRight: <span>TopBarRight...</span>,
    views,
  },
};
