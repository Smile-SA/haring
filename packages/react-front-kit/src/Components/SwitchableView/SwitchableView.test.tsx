import { renderWithProviders } from '@smile/react-front-kit-shared/test-utils';

import { SwitchableView } from './SwitchableView';
import { viewsMock } from './SwitchableView.mock';

describe('SwitchableView', () => {
  beforeEach(() => {
    // Prevent mantine random ID
    Math.random = () => 0.42;
  });
  it('matches snapshot', () => {
    const { container } = renderWithProviders(
      <SwitchableView views={viewsMock} />,
    );
    expect(container).toMatchSnapshot();
  });
});
