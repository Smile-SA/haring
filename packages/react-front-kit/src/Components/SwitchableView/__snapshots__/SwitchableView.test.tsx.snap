// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchableView matches snapshot 1`] = `
<div>
  <div
    class="mantine-Paper-root mantine-z9bo1e"
  >
    <div
      class="mantine-v9qkjz"
    >
      <span
        class="mantine-qobh7"
      >
        <div
          class="mantine-SegmentedControl-root mantine-np8w2m"
        >
          <span
            class="mantine-SegmentedControl-indicator mantine-1dj3pci"
          />
          <div
            class="___ref-control mantine-SegmentedControl-control mantine-SegmentedControl-controlActive mantine-92jry0"
          >
            <input
              checked=""
              class="mantine-1qq2b01 mantine-SegmentedControl-input"
              id="mantine-f4bipx4bi-a"
              name="mantine-f4bipx4bi"
              type="radio"
              value="a"
            />
            <label
              class="___ref-label mantine-SegmentedControl-label mantine-1sk7jdh"
              data-active="true"
              for="mantine-f4bipx4bi-a"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M80,64a8,8,0,0,1,8-8H216a8,8,0,0,1,0,16H88A8,8,0,0,1,80,64Zm136,56H88a8,8,0,0,0,0,16H216a8,8,0,0,0,0-16Zm0,64H88a8,8,0,0,0,0,16H216a8,8,0,0,0,0-16ZM44,52A12,12,0,1,0,56,64,12,12,0,0,0,44,52Zm0,64a12,12,0,1,0,12,12A12,12,0,0,0,44,116Zm0,64a12,12,0,1,0,12,12A12,12,0,0,0,44,180Z"
                />
              </svg>
            </label>
          </div>
          <div
            class="___ref-control mantine-1ryipl8 mantine-SegmentedControl-control"
          >
            <input
              class="mantine-1qq2b01 mantine-SegmentedControl-input"
              id="mantine-f4bipx4bi-b"
              name="mantine-f4bipx4bi"
              type="radio"
              value="b"
            />
            <label
              class="___ref-label mantine-SegmentedControl-label mantine-1sk7jdh"
              for="mantine-f4bipx4bi-b"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M104,40H56A16,16,0,0,0,40,56v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V56A16,16,0,0,0,104,40Zm0,64H56V56h48v48Zm96-64H152a16,16,0,0,0-16,16v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V56A16,16,0,0,0,200,40Zm0,64H152V56h48v48Zm-96,32H56a16,16,0,0,0-16,16v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V152A16,16,0,0,0,104,136Zm0,64H56V152h48v48Zm96-64H152a16,16,0,0,0-16,16v48a16,16,0,0,0,16,16h48a16,16,0,0,0,16-16V152A16,16,0,0,0,200,136Zm0,64H152V152h48v48Z"
                />
              </svg>
            </label>
          </div>
          <div
            class="___ref-control mantine-1ryipl8 mantine-SegmentedControl-control"
          >
            <input
              class="mantine-1qq2b01 mantine-SegmentedControl-input"
              id="mantine-f4bipx4bi-c"
              name="mantine-f4bipx4bi"
              type="radio"
              value="c"
            />
            <label
              class="___ref-label mantine-SegmentedControl-label mantine-1sk7jdh"
              for="mantine-f4bipx4bi-c"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M239.2,97.29a16,16,0,0,0-13.81-11L166,81.17,142.72,25.81h0a15.95,15.95,0,0,0-29.44,0L90.07,81.17,30.61,86.32a16,16,0,0,0-9.11,28.06L66.61,153.8,53.09,212.34a16,16,0,0,0,23.84,17.34l51-31,51.11,31a16,16,0,0,0,23.84-17.34l-13.51-58.6,45.1-39.36A16,16,0,0,0,239.2,97.29Zm-15.22,5-45.1,39.36a16,16,0,0,0-5.08,15.71L187.35,216v0l-51.07-31a15.9,15.9,0,0,0-16.54,0l-51,31h0L82.2,157.4a16,16,0,0,0-5.08-15.71L32,102.35a.37.37,0,0,1,0-.09l59.44-5.14a16,16,0,0,0,13.35-9.75L128,32.08l23.2,55.29a16,16,0,0,0,13.35,9.75L224,102.26S224,102.32,224,102.33Z"
                />
              </svg>
            </label>
          </div>
        </div>
      </span>
    </div>
    <div
      class="mantine-l6ko5p"
    >
      <div
        class="mantine-Center-root mantine-ojrz4j"
        style="background-color: lightgray; height: 200px;"
      >
        View A
      </div>
    </div>
  </div>
</div>
`;
